File: packages/gbnf-experiment/data/20260909T124157Z_4ee67ba2/ported_implementation/tests/unit/rules-builder/parse-char.test.ts
import assert from 'node:assert/strict';
import { describe, test } from 'node:test';
import { parseChar } from '../../../src/rules-builder/parse-char.ts';
import { GrammarParseError } from '../../../src/utils/errors/index.ts';
import { ord } from '../../helpers.ts';

const PREFIX_LENGTH = 'root ::= "'.length;

const SIMPLE: [string, string, number][] = [
  ['escaped 8-bit unicode char', 'a', ord('a')],
  ['escaped 8-bit unicode char', '9', ord('9')],
];

const COMPLEX: [string, string, number, number][] = [
  ['escaped 8-bit unicode char', '\\x2A', ord('\x2A'), 4],
  ['escaped 16-bit unicode char', '\\u006F', ord('o'), 6],
  ['escaped 32-bit unicode char', '\\U0001F4A9', 128169, 10],
  ['escaped tab char', '\\t', ord('\t'), 2],
  ['escaped new line char', '\\n', ord('\n'), 2],
  ['escaped \r char', '\\r', ord('\r'), 2],
  ['escaped quote char', '\\"', ord('"'), 2],
  ['escaped [ char', '\\[', ord('['), 2],
  ['escaped ] char', '\\]', ord(']'), 2],
  ['escaped \\ char', '\\\\', ord('\\'), 2],
];

const RAISES: [string, number][] = [
  ['', 0],
  ['a', 1],
  ['a', 2],
];

describe('parse char', () => {
  for (const [index, [description, char, codePoint]] of SIMPLE.entries()) {
    test(`simple: ${description} [${index}]`, () => {
      const grammar = `root ::= "${char}" "foo"`;
      assert.deepStrictEqual(parseChar(grammar, PREFIX_LENGTH), [codePoint, 1]);
    });
  }

  for (const [description, escapedChar, codePoint, incPos] of COMPLEX) {
    test(`complex: ${description}`, () => {
      const grammar = `root ::= "${escapedChar}" "foo"`;
      assert.deepStrictEqual(parseChar(grammar, PREFIX_LENGTH), [codePoint, incPos]);
    });
  }

  for (const [input, pos] of RAISES) {
    test(`raises for ${JSON.stringify(input)} @ ${pos}`, () => {
      assert.throws(() => parseChar(input, pos), GrammarParseError);
    });
  }
});

File: proposed-projects/round-trip-experiment/reverse/20260909T002206Z_273cf6b9/20260910T042804Z_e72f6c04/ported_implementation/gbnf/rules_builder/parse_char_test.py
import pytest

from ..utils.errors import GrammarParseError
from .parse_char import parse_char

PREFIX_LENGTH = len('root ::= "')


@pytest.mark.parametrize(
    ("description", "char", "code_point"),
    [
        ("escaped 8-bit unicode char", "a", ord("a")),
        ("escaped 8-bit unicode char", "9", ord("9")),
    ],
)
def test_simple(description, char, code_point):
    grammar = f'root ::= "{char}" "foo"'
    assert parse_char(grammar, PREFIX_LENGTH) == (code_point, 1)


@pytest.mark.parametrize(
    ("description", "escaped_char", "code_point", "inc_pos"),
    [
        ("escaped 8-bit unicode char", "\\x2A", 0x2A, 4),
        ("escaped 16-bit unicode char", "\\u006F", 0x6F, 6),
        ("escaped 32-bit unicode char", "\\U0001F4A9", 128169, 10),
        ("escaped tab char", "\\t", 9, 2),
        ("escaped new line char", "\\n", 10, 2),
        ("escaped \r char", "\\r", 13, 2),
        ("escaped quote char", '\\"', 34, 2),
        ("escaped [ char", "\\[", 91, 2),
        ("escaped ] char", "\\]", 93, 2),
        ("escaped \\ char", "\\\\", 92, 2),
    ],
)
def test_complex(description, escaped_char, code_point, inc_pos):
    grammar = f'root ::= "{escaped_char}" "foo"'
    assert parse_char(grammar, PREFIX_LENGTH) == (code_point, inc_pos)


@pytest.mark.parametrize(
    ("input", "pos"),
    [
        ("", 0),
        ("a", 1),
        ("a", 2),
    ],
)
def test_it_raises(input, pos):
    with pytest.raises(GrammarParseError):
        parse_char(input, pos)
